akka-http/src/main/scala/akka/http/scaladsl/server/RequestContextImpl.scala

```diff
@@ -23,14 +23,15 @@ import scala.concurrent.{ ExecutionContextExecutor, Future }
 private[http] class RequestContextImpl(
   val request:          HttpRequest,
   val unmatchedPath:    Uri.Path,
+  val unusedParameters: Set[String],
   val executionContext: ExecutionContextExecutor,
   val materializer:     Materializer,
   val log:              LoggingAdapter,
   val settings:         RoutingSettings,
   val parserSettings:   ParserSettings) extends RequestContext {
 
   def this(request: HttpRequest, log: LoggingAdapter, settings: RoutingSettings, parserSettings: ParserSettings)(implicit ec: ExecutionContextExecutor, materializer: Materializer) =
-    this(request, request.uri.path, ec, materializer, log, settings, parserSettings)
+    this(request, request.uri.path, request.uri.query().toMap.keySet, ec, materializer, log, settings, parserSettings)
```

Comment:
I'm not sure we would want to do this here: we have made the parsing of the query string as a set of name-value parameters lazy (they are not parsed until you actually use them as name-value pairs), because it is not always needed (so postponing parsing them can be good for performance), and may even fail: there are query strings that are allowed by the URI spec, but cannot be parsed into name-value pairs.

Reply:
What is your suggestion?

Reply:
In the other hand, every time a call to `parameter(...)` or `parameters(...)` the following method is executed
```scala
Try(request.uri.query())
```

which does not seem very efficient either. Maybe just getting all the query parameters in the `RequestContext` (lazy if you want) and then having a set with the currently visited would help with this.